src/Catalyst.Abstractions/Kvm/Models/BlockParameters.cs

```diff
@@ -1,84 +1,112 @@
 
         public BlockParameterType Type { get; set; }
-        public long? BlockNumber { get; set; }
+        public long? BlockNumber { get; }
+        
+        public Keccak BlockHash { get; }
+
+        public bool RequireCanonical { get; }
 
         public BlockParameter() { }
 
         public BlockParameter(BlockParameterType type)
         {
             Type = type;
-            BlockNumber = null;
         }
 
-        public void FromJson(string jsonValue)
+        public BlockParameter(long number)
+        {
+            Type = BlockParameterType.BlockNumber;
+            BlockNumber = number;
+        }
+        
+        public BlockParameter(Keccak blockHash)
+        {
+            Type = BlockParameterType.BlockHash;
+            BlockHash = blockHash;
+        }
+        
+        public BlockParameter(Keccak blockHash, bool requireCanonical)
+        {
+            Type = BlockParameterType.BlockHash;
+            BlockHash = blockHash;
+            RequireCanonical = requireCanonical;
+        }
+        
+        public static BlockParameter FromJson(string jsonValue)
         {
             switch (jsonValue)
             {
                 case { } earliest when string.Equals(earliest, "earliest", StringComparison.InvariantCultureIgnoreCase):
-                    Type = BlockParameterType.Earliest;
-                    return;
+                    return Earliest;
                 case { } pending when string.Equals(pending, "pending", StringComparison.InvariantCultureIgnoreCase):
-                    Type = BlockParameterType.Pending;
-                    return;
+                    return Pending;
                 case { } latest when string.Equals(latest, "latest", StringComparison.InvariantCultureIgnoreCase):
-                    Type = BlockParameterType.Latest;
-                    return;
+                    return Latest;
                 case { } empty when string.IsNullOrWhiteSpace(empty):
-                    Type = BlockParameterType.Latest;
-                    return;
+                    return Latest;
                 case null:
-                    Type = BlockParameterType.Latest;
-                    return;
+                    return Latest;
+                case { } hash when hash.Length == 66 && hash.StartsWith("0x"):
+                    return Latest;
                 default:
-                    Type = BlockParameterType.BlockNumber;
-                    BlockNumber = LongConverter.FromString(jsonValue.Trim('"'));
-                    return;
+                    return new BlockParameter(LongConverter.FromString(jsonValue.Trim('"')));
             }
         }
 
         public override string ToString()
         {
-            return $"{Type}, {BlockNumber}";
+            return $"{Type}, {BlockNumber?.ToString() ?? BlockHash?.ToString()}";
+        }
+        
+        public bool Equals(BlockParameter other)
+        {
+            if (ReferenceEquals(null, other)) return false;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=4979613)